server/consumer.go

```diff
@@ -1031,47 +1051,174 @@ func (o *Consumer) needAck(sseq uint64) bool {
+type waitQueue struct {
+	rp, wp int
+	reqs   []*waitingRequest
+}
+
+// Create a new ring buffer with at most max items.
+func newWaitQueue(max int) *waitQueue {
+	return &waitQueue{rp: -1, reqs: make([]*waitingRequest, max)}
+}
+
+var (
+	errWaitQueueFull = errors.New("wait queue is full")
+	errWaitQueueNil  = errors.New("wait queue is nil")
+)
+
+// Adds in a new request.
+func (wq *waitQueue) add(req *waitingRequest) error {
+	if wq == nil {
+		return errWaitQueueNil
+	}
+	if wq.isFull() {
+		return errWaitQueueFull
+	}
+	wq.reqs[wq.wp] = req
+	// TODO(dlc) - Could make pow2 and get rid of mod.
+	wq.wp = (wq.wp + 1) % cap(wq.reqs)
+
+	// Adjust read pointer if we were empty.
+	if wq.rp < 0 {
+		wq.rp = 0
+	}
+
+	return nil
+}
+
+func (wq *waitQueue) isFull() bool {
+	return wq.rp == wq.wp
+}
+
+func (wq *waitQueue) len() int {
+	if wq == nil || wq.rp < 0 {
+		return 0
+	}
+	if wq.rp < wq.wp {
+		return wq.wp - wq.rp
+	}
+	return cap(wq.reqs) - wq.rp + wq.wp
+}
+
+// Peek will return the next request waiting or nil if empty.
+func (wq *waitQueue) peek() *waitingRequest {
+	if wq == nil {
+		return nil
+	}
+	var wr *waitingRequest
+	if wq.rp >= 0 {
+		wr = wq.reqs[wq.rp]
+	}
+	return wr
+}
+
+// pop will return the next request and move the read cursor.
+func (wq *waitQueue) pop() *waitingRequest {
+	wr := wq.peek()
+	if wr != nil {
+		wr.n--
+		if wr.n <= 0 {
+			wq.reqs[wq.rp] = nil
+			wq.rp = (wq.rp + 1) % cap(wq.reqs)
+			// Check if we are empty.
+			if wq.rp == wq.wp {
+				wq.rp, wq.wp = -1, 0
+			}
 		}
 	}
-	return bs
+	return wr
 }
 
 // processNextMsgReq will process a request for the next message available. A nil message payload means deliver
 // a single message. If the payload is a number parseable with Atoi(), then we will send a batch of messages without
 // requiring another request to this endpoint, or an ACK.
-func (o *Consumer) processNextMsgReq(_ *subscription, _ *client, _, reply string, msg []byte) {
-	// Check payload here to see if they sent in batch size.
-	batchSize := batchSizeFromMsg(msg)
-
+func (o *Consumer) processNextMsgReq(_ *subscription, c *client, _, reply string, msg []byte) {
 	o.mu.Lock()
 	mset := o.mset
 	if mset == nil || o.isPushMode() {
 		o.mu.Unlock()
 		return
 	}
-	shouldSignal := false
+
+	sendErr := func(status int, description string) {
+		sendq := mset.sendq
+		o.mu.Unlock()
+		hdr := []byte(fmt.Sprintf("NATS/1.0 %d %s\r\n\r\n", status, description))
+		pmsg := &jsPubMsg{reply, reply, _EMPTY_, hdr, nil, nil, 0}
+		sendq <- pmsg // Send message.
+	}
+
+	if o.waiting.isFull() {
+		// If our waiting queue is full return an empty response with the proper header.
+		// FIXME(dlc) - Should we do advisory here as well?
+		sendErr(500, "WaitQueue Exceeded")
+		return
+	}
+
+	// Check payload here to see if they sent in batch size or a formal request.
+	expires, batchSize, noWait, err := nextReqFromMsg(msg)
+	if err != nil {
+		sendErr(400, "Bad Request")
```

Comment:
perhaps an opportunity to make the reason for failure clearer here? Request JSON invalid or invalid batch size number or similar?

Reply:
Possibly, although that is the server spending time and we really do not know much more than that. Could distinguish a bit more but not sure its worth it.

Reply:
It's not a lot of extra work - we already return a err there, sute its a bit no question but it would be a big help for users when debugging things.

Reply:
We only have two options, could not parse JSON or raw batchSize number..

Reply:
Yup. So should be easy to add user friendly behaviour :) it’s not like users know we have only these 2 without reading code